Complete update to ZF v3 releases

weierophinney opened the pull request:
This patch builds on #346, to provide forwards compatibility with v3 releases of ZF components.

In addition to the changes already present in #346, this patch:

- Moves all factories originally defined in the `Module::getServiceConfig()` and `Module::getControllerConfig()` methods into their own classes, all under the `ZF\Apigility\Admin\Factory` namespace.
- Redefines all invokables as factories using `InvokableFactory`; aliases were also created when necessary.
- Updates configuration to use `::class` notation.
- Updates all existing factories to be forwards compatible with zend-servicemanager v3.
- Ensures all listeners are attached in a forwards compatible fashion.
- Ensures all triggers happen correctly.
- Ensures checks for `RouteMatch` will check for either the v2 or v3 router releases.
- Updates to zf-apigility 1.3 stable release.
- Provides an upgrade script, `vendor/bin/apigility-update-to-1.5.php`, which will automate the process of upgrading to ZF v3 releases within an existing Apigility application.

~~Currently, tests pass with v2 ZF releases, but fail with dozens of errors and several failures when tested against v3 releases; this will be the focus for completion of the patch.~~

### TODO

- [x] Extract listeners from the `Module` class to their own classes (and include tests):
  - [x] normalizeMatchedControllerServiceName
  - [x] normalizeMatchedInputFilterName
  - [x] onRoute
  - [x] onRender
  - [x] onFinish
  - [x] HalPlugin::onRenderEntity
  - [x] HalPlugin::onRenderCollectionEntity
- [x] Test all new factories
  - [x] AuthenticationControllerFactory
  - [x] AuthenticationModelFactory
  - [x] AuthorizationControllerFactory
  - [x] AuthorizationModelFactory
  - [x] ConfigControllerFactory
  - [x] ContentNegotiationModelFactory
  - [x] ContentNegotiationResourceFactory
  - [x] DbAdapterModelFactory
  - [x] DbAdapterResourceFactory
  - [x] DbAutodiscoveryModelFactory
  - [x] DoctrineAdapterModelFactory
  - [x] DoctrineAdapterResourceFactory
  - [x] ModuleConfigControllerFactory
  - [x] ModuleCreationControllerFactory
  - [x] ModuleModelFactory
  - [x] ModulePathSpecFactory
  - [x] ModuleResourceFactory
  - [x] ModuleVersioningModelFactoryFactory
  - [x] RestServiceModelFactoryFactory
  - [x] RestServiceResourceFactory
  - [x] RpcServiceModelFactoryFactory
  - [x] RpcServiceResourceFactory
  - [x] SourceControllerFactory
  - [x] VersioningControllerFactory
  - [x] VersioningModelFactoryFactory

weierophinney commented:
Ready for review!

Ping @zfcampus/community-review-team and @webimpress

weierophinney commented:
... and tests are passing, after doing a re-install to fix a `composer.lock` issue. REALLY READY NOW! :smile:

webimpress commented:
https://github.com/weierophinney/zf-apigility-admin/blob/ae607978c028bd3f468a9043483bcef96ed798ac/config/module.config.php#L385

missing `C`

I'll review it in full tomorrow.
It's was a lot of work, thanks @weierophinney ! 😄

weierophinney commented:
> missing C

Fixed - thanks, @webimpress !

webimpress commented on src/Factory/AuthenticationControllerFactory.php:
```diff
@@ -0,0 +1,42 @@
+     * @param null|string $requestedName
+     * @return AuthenticationController
+     */
+    public function createService(ServiceLocatorInterface $container, $cName = null, $requestedName = null)
```
Why in this factory we have two extra params `$cName` and `$requestedName`? Do I miss something?

weierophinney replied:
It's required here as there are several aliases (one each for HTTP basic, HTTP digest, and oauth2) that resolve to this factory, and construction will occur differently based on the alias requested. In zend-servicemanager v2, it passes the requested name and normalized name as arguments, but they're not strictly part of the interface; this simply makes use of that. (The original factory did as well.)

weierophinney replied:
And yet... the factory isn't making use of it. Let me double check.

weierophinney replied:
I'll remove this; it's not being used.

webimpress commented on composer.json:
```diff
@@ -16,57 +16,60 @@
     "extra": {
         "branch-alias": {
             "dev-master": "1.4-dev",
             "dev-develop": "1.5-dev"
```
it should be updated (will be done after merge?)

weierophinney replied:
That gets updated after the merge to master.

webimpress commented on src/Factory/DoctrineAdapterResourceFactory.php:
```diff
@@ -0,0 +1,37 @@
+     */
+    public function __invoke(ContainerInterface $container)
+    {
+        if (! $services->has(DoctrineAdapterModel::class)) {
```
`$container` please, not `$services`
(and the same twice below in this method)

webimpress commented on src/Model/DoctrineAdapterResource.php:
```diff
@@ -7,65 +7,41 @@
      * @param DoctrineAdapterModel $model
      */
-    public function __construct(DoctrineAdapterModel $model)
+    public function __construct(DoctrineAdapterModel $model, array $loadedModules)
```
Is it gonna be BC break? Maybe it should have default value - an empty array?
Please update also php-docs (missing 2nd param)

weierophinney replied:
Shouldn't be an issue, as you'll most likely be using the factory we provide, but I'll update regardless.

webimpress commented on src/Factory/DoctrineAdapterModelFactory.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    public function __invoke(ContainerInterface $container)
+    {
+        if (! $services->has('config')) {
```
`$container` please (also twice below in this method)

webimpress commented on src/Factory/ModuleModelFactory.php:
```diff
@@ -0,0 +1,57 @@
+        if (! $container->has('ModuleManager')) {
+            throw new ServiceNotCreatedException(sprintf(
+                'Cannot create %s service because ModuleManager service is not present',
+                Model\ModuleModel::class
```
without `Model\`, just `ModuleModel::class`

webimpress commented on src/Factory/RestServiceModelFactoryFactory.php:
```diff
@@ -0,0 +1,78 @@
+use Interop\Container\ContainerInterface;
+use Zend\EventManager\SharedEventManagerInterface;
+use Zend\ServiceManager\Exception\ServiceNotCreatedException;
+use ZF\Apigility\Admin\Model\DbConnectedRestServiceModel;
```
It seems to be not used in this class

weierophinney replied:
Actually, indicative of another error; the first `$sharedEvents->attach()` should be using this class! Updating now.

webimpress commented on src/Model/AbstractPluginManagerModel.php:
```diff
@@ -51,18 +53,59 @@ protected function getPlugins()
+     */
+    private function filterPluginName($name)
+    {
+        return ! (bool) preg_match('/^zend(hydrator|stdlib)/', $name);
```
useless type casting here
http://php.net/manual/en/function.preg-match.php#refsect1-function.preg-match-returnvalues

weierophinney replied:
We want to ensure the method returns a boolean, and `preg_match()` returns an int. If we later change the internals to do something else, this ensures the signature and usage remains the same.

webimpress commented on src/Factory/RestServiceModelFactoryFactory.php:
```diff
@@ -0,0 +1,78 @@
+        $sharedEvents = $container->get('SharedEventManager');
+        $this->attachSharedListeners($sharedEvents, $container);
+
+        return new Model\RestServiceModelFactory(
```
without `Model\` here please

webimpress commented on src/Model/AbstractAutodiscoveryModel.php:
```diff
@@ -60,6 +65,10 @@
     public function getServiceLocator()
     {
+        if (! $this->serviceLocator) {
+            throw new Exception('The AbstractAutodiscoveryModel must be composed with a service locator');
```
missing `@throws` in php-docs now

weierophinney replied:
Added now.

webimpress commented on src/Model/DocumentationModelFactory.php:
```diff
@@ -1,29 +1,53 @@
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
     {
-        if (! $services->has('ZF\Configuration\ConfigResourceFactory')) {
+        if (! $services->has(ConfigResourceFactory::class)) {
```
`$container` please (two more below as well)

webimpress commented on src/Module.php:
```diff
@@ -0,0 +1,598 @@
+     */
+    protected function injectServiceLinks(Entity $halEntity, HalJsonModel $model, $e)
+    {
+        $entity = $halEntity->entity;
```
`->getEntity()` please (zf-hal 1.4)

webimpress commented on src/Module.php:
```diff
@@ -0,0 +1,598 @@
+    public function onRenderEntity($e)
+    {
+        $halEntity = $e->getParam('entity');
+        $entity    = $halEntity->entity;
```
`->getEntity()` please (zf-hal 1.4)

weierophinney commented:
@webimpress All listeners are now extracted and tested, including updates to use `ZF\Hal\Entity::getEntity()` instead of `ZF\Hal\Entity::$entity`. Turning to the factories now.

weierophinney commented:
All factories now have test cases; doing so unearthed a few issues, which I was able to fix.

Additionally, @webimpress discovered an issue with how `InjectModuleResourceLinksListener` was attaching its `onHalRenderEvents` listener. The original code was passing an array events to the event manager's `attach()` method; however, this is no longer supported in zend-eventmanager v3. I've updated the tests for that listener and the code for it to attach the listener to each event individually.

weierophinney commented:
Ping @zfcampus/community-review-team ; ready for review (@webimpress is already working on it...).

webimpress commented on src/Factory/DbAutodiscoveryModelFactory.php:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        $instance = new DbAutodiscoveryModel($container->get('config'));
+        $instance->setServiceLocator($container);
```
As you pointed in comment https://github.com/zfcampus/zf-apigility-admin/pull/348/files#diff-c7e9fdf4a6c840f9dd80cd5050433a48R10 we need it here.

But should we also update [`AbstractAutodiscoveryModel::setServiceLocator`](https://github.com/weierophinney/zf-apigility-admin/blob/feature/346/src/Model/AbstractAutodiscoveryModel.php#L82) to accept `ContainerInterface` instead of `ServiceLocatorInterface`?

weierophinney replied:
I thought about that yesterday as I was writing the tests for this factory. However, changing the type accepted would break BC for any extending classes that override the method. There likely aren't any, but I'd rather err on the side of caution.

webimpress commented on src/Model/AbstractAutodiscoveryModel.php:
```diff
@@ -3,10 +3,15 @@
+/**
+ * This class is instantiated with a $config in some implementations (DbAutodiscoveryModel)
+ * but this is dependent on the root service locator for the moduleHasService call below
+ * and that must be injected into any class extending this abstract.
```
Is it then BC Break? I think it should be pointed then in Release Notes.

weierophinney replied:
No; removing an interface being implemented, but keeping the methods, does not pose a BC break. I'll still message it in the notes, however, as it *is* a change.

webimpress commented:
A few thoughts:

- [ ] Move factories to corresponding namespaces - now there is mess, some factories are in `Factories` and some in `Controller` namespace, etc. Not sure if new `Factory` namespace is needed as long as all factories are with `Factory` suffix.

- [ ] `s/hal/Hal/` in [`Controller\AuthenticationController#L69`](https://github.com/weierophinney/zf-apigility-admin/blob/feature/346/src/Controller/AuthenticationController.php#L69)

- [ ] [`controllers.aliases` in `module..config.php`](https://github.com/weierophinney/zf-apigility-admin/blob/feature/346/config/module.config.php#L50-L66) - do we need them? Can we sort it out a bit? - Not sure if everything is right in `controllers.factories` - for example:
`Controller\Validators::class => Controller\ValidatorsControllerFactory::class`,
but class `Controller\Validators` doesn't exist (I know, `::class` works anyway), but maybe we should have in `aliases`:
`Controller\Validators::class => Controller\ValidatorController::class`
and in `factories`:
`Controller\ValidatorsController::class => Controller\ValidatorsControllerFactory::class`

- [ ] Alphabetical order in:
   - [ ] `service_manager.factories`
   - [ ] `controllers.aliases`
   - [ ] `controller.factories`
   - [ ] `input_filters.aliases`
   - [ ] `input_filters.factories`

- [ ] Short array notation in files `test/Model/TestAsset/module/Version/config/*.config.php.dist` and remove `controllers.invokables` in `module.config.php.dist`

- [ ] `RpcServiceResource::__construct` and `RestServiceResource::getModel` php-docs needs update

@weierophinney Thanks for your work!
I'll continue my review later.
I'd like to run apigility admin on my local and check if everything works as expected. Will do it later.

weierophinney commented:
> controllers.aliases in module..config.php - do we need them?

Yes, they're needed, for a few reasons.

Every single alias I created was originally defined as an invokable; in each of these cases, though, the service name did not match the actual class being created, but instead truncated the class name to omit the `Controller` suffix.

Internally, when we use the `invokables` key, it does exactly what I did here: define an alias of the short name to the FQCN, and a factory mapping the FQCN to `InvokableFactory`.

In the cases where we have a short name pointing to a factory name, these generally already existed, and modifying them to do create alias/FQCN + FQCN/factory pairs would potentially break cases where an alternate factory was defined by an end-user.

weierophinney commented:
Factories are all moved into their respective namespaces now, @webimpress .

weierophinney commented:
I'm now testing this against the skeleton, and have discovered that:

- `StrategyController` is calling `getServiceLocator()`. It may have an actual need to do so, but I'll update it to receive it during instantiation.
- Some of the existing controller factories were written as invokables, and thus call `getServiceLocator()` internally in order to retrieve the parent context. I will update these to implement `FactoryInterface`.

weierophinney commented:
All controller factories are now updated, and the `StrategyController` was also updated to accept a `ContainerInterface` to the constructor.

webimpress commented on test/Model/RpcServiceModelFactoryFactoryTest.php:
```diff
@@ -0,0 +1,104 @@
+        $this->container->get('SharedEventManager')->willReturn($sharedEvents);
+        $this->container->get(ModuleModel::class)->willReturn($moduleModel);
+
+        $restFactory = $factory($this->container->reveal());
```
It should be `$rpcFactory`

weierophinney commented:
I've rebased this against #350 at this time.

webimpress commented:
@weierophinney 
this part is gone after rebase:
```php
 -    public function init(ModuleManagerInterface $modules)
 -    {
 -        $loaded = $modules->getLoadedModules(false);
 -        if (isset($loaded['ZF\Apigility\Admin\Ui'])) {
 -            return;
 -        }
 -
 -        $modules->loadModule('ZF\Apigility\Admin\Ui');
 -    }
```
Now we need add module `'ZF\\Apigility\\Admin\\Ui'` manually into `development.config.php`, is it fine?

weierophinney commented:
> Now we need add module 'ZF\\Apigility\\Admin\\Ui' manually into development.config.php, is it fine?

Nope, bad rebase. I'll push that back in a moment.

webimpress commented on src/InputFilter/Factory/InputFilterInputFilterFactory.php:
```diff
@@ -14,19 +15,36 @@
+     */
+    public function createService(ServiceLocatorInterface $container)
+    {
+        if ($container instanceof AbstractPluginManager) {
```
`AbstractPluginManager` is missing. Needed import:
`use Zend\ServiceManager\AbstractPluginManager;`

It seems to be broken after rebase, it works fine before.

weierophinney replied:
Fixed.

webimpress commented on src/Controller/DbAutodiscoveryControllerFactory.php:
```diff
@@ -1,32 +1,45 @@
+        if ($container instanceof AbstractPluginManager) {
+            $container = $container->getServiceLocator() ?: $container;
+        }
+        return $this($container, DbAutodiscoverController::class);
```
typo, missing `y`- should be `DbAutodiscoveryController`

webimpress commented on src/Listener/InjectModuleResourceLinksListenerFactory.php:
```diff
@@ -0,0 +1,22 @@
+class InjectModuleResourceLinksListenerFactory
+{
+    /**
+     * @param ContainerInterface $container
```
missing `@return` tag

weierophinney replied:
Fixed.

webimpress commented on src/Model/RestServiceModelFactoryFactory.php:
```diff
@@ -0,0 +1,73 @@
+            RestServiceModel::class,
+            'fetch',
+            [DoctrineRestServiceModel::class, 'onFetch']
+        );
```
hm... can it be moved into `zf-apigility-doctrine` module ?

weierophinney replied:
It could likely be moved to a delegator in that module. However, since we'll be having this released first, it would break for existing users of that module.

We'll have to coordinate releases at a later point.

webimpress commented on src/Module.php:
```diff
@@ -0,0 +1,150 @@
+use ZF\Hal\Link\Link;
+use ZF\Hal\Link\LinkCollection;
+use ZF\Hal\Entity;
+use ZF\Hal\View\HalJsonModel;
```
Most of them are not used, please clear it

weierophinney replied:
Done!

weierophinney commented:
@webimpress Just pushed commits addressing your last round of feedback.

weierophinney commented:
I've just done the following to test that upgrading works as expected:

- Installed a fresh Apigility project, using the installer from the website.
- Updated the `composer.json` to make the following changes:
  - Added a path repository pointing to my local checkout of zf-apigility-admin
  - Updated the zf-apigility-admin constraint to read `@dev`
- Ran `composer update`

This picked up the latest releases of all Apigility components, with the exception of zf-development-mode (as the current default constraint is `~2.0`), and the latest v2 releases of all zf components (as the default skeleton has zendframework with the constraint `^2.5.3` as the default).

This *mostly* works; the one issue is that we've removed rwoverdijk/assetmanager as a requirement in zf-apigility, zf-apigility-admin, and zf-apigility-admin-ui, but the skeleton enables the `AssetManager` module. This can be corrected by running:

```bash
$ composer require rwoverdijk/assetmanager
```

after an update, however.

Once I'd done that, I tested all functionality of the admin UI, and found everything worked perfectly out-of-the-box, with the exception of the UI issues that @webimpress and I discussed over the weekend (in particular, the sidebar does not refresh correctly after creating services; additionally, the fully service name is listed in the service title — e.g., `Test-V1-Rest-Foo-Controller` vs the simpler `Foo`).

My thoughts on this are that we will need to document the following migration steps:

- Initial update will be:
  - `composer update`
  - `composer require rwoverdijk/assetmanager`
- More comprehensive update will be:
  - `composer.json` changes:
    - Remove the entry for `zendframework/zendframework` from the requirements.
    - Update the `zfcampus/zf-development-mode` constraint to `^2.1 || ^3.0` (or just `^3.0`).
    - Remove the entry for `rwoverdijk/assetmanager`.
    - Add `zfcampus/zf-asset-manager` with constraint `^1.0` as a development requirement.
  - `config/modules.config.php` changes:
    - Remove the following entries:
      - `AssetManager`
      - `ZF\DevelopmentMode`
    - Add the following entries to `config/modules.config.php` (at the top of the list):
      - `Zend\Db`
      - `Zend\Filter`
      - `Zend\Hydrator`
      - `Zend\InputFilter`
      - `Zend\Paginator`
      - `Zend\Router`
      - `Zend\Validator`
  - Remove, recursively, the `vendor/` directory and `composer.lock`.
  - Run `composer install`.

The above will update the application to the latest versions of all Apigility modules, as well as ZF v3 releases.

weierophinney commented:
@ezimuel has suggested we provide a vendor binary script that performs the migration steps; I'm going to work on that now.

weierophinney commented:
Upgrade script is now provided!

weierophinney commented:
CHANGELOG written, and migration instructions provided in the README.

webimpress commented:
@weierophinney Please have a look on my PR with some small fixes to your branch: https://github.com/weierophinney/zf-apigility-admin/pull/1
I've added also some TODO / Questions in the description, also copied below (some of them we discussed yesterday):

**TODO / Quesions:**
- [ ] `setRequest` method in `AuthorizationController` should be removed or at least marked as `@deprecated`
- [ ] `VersioningModelFactoryFactory` should be marked as `@deprecated` as long as `VersioningModelFactory` is deprecated
- [ ] upgrade script should leave one empty line at the end of `modules.config.php` and `composer.json`
- [ ] upgrade script should remove `ZFTool` module from `development.config.php` and `development.config.php.dist`
- [ ] HHVM tests need to be checked - many of them are failing
- [ ] remove `markTestSkipped` in some mongo tests (fix if needed) - all test should run on travis
- [ ] check tests `sh: 1: tree: not found` in https://travis-ci.org/zfcampus/zf-apigility-admin/jobs/150746938#L415

Thanks!

webimpress commented:
@weierophinney upgrade script does not remove ZFTool from modules list (`development.config.php` and `development.config.php.dist`).
Now one test is failing on PHP7. I would go with slightly different travis configuration for mongo:
https://github.com/webimpress/zf-apigility-doctrine/blob/0294a485c27d6fa3564f2d1dba1759b622592599/.travis.yml
I'm not sure about HHVM and mongo configuration, it doesn't work for me now. But maybe we can skip it for now on hhvm?

webimpress commented on bin/apigility-upgrade-to-1.5:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    // Update existing entries
+    $composer['require']['zfcampus/zf-development-mode'] = '^2.1 || ^3.0';
```
@weierophinney  as we discuss yesterday, it should be changed to `^3.0` only

weierophinney replied:
Done.
